DRAFT: Add support for Deye CAN PCS batteries

GHswitt opened the pull request:
This adds support for Deye batteries using the CAN PCS port of the battery.
Via the PCS port the battery sends out the overall status and status frames for each installed module. This commit was tested with a battery with two Deye AI-W5.1 modules.

bms_combine_DEYE-PCS_CAN.yaml:    Combination of the BMS parts
bms_can_DEYE_PCS.yaml:            Main Deye BMS data
bms_can_DEYE_PCS_Module.yaml:     Deye battery module data.
                                  This packages needs to be included
                                  multiple times. Once for each module.
bms_errors_bitmask_DEYE_CAN.yaml: Error mapping Deye -> yambms.
one-bms_DEYE-PCS_CAN.yaml:        Example test file.

Sleeper85 commented:
Regarding the word `PCS`, this stands for `Personal Computer Software`.
I'll try not to use it everywhere except for the YAML that collects information from the PCS port.

The PCS port provides aggregate information for all modules (much like YamBMS does during the combination process), but this information is not useful.

The InterCAN (IN/OUT port) network provides information for all modules, I believe this is the only necessary source of information.

Importing modules into the main YAML would be done like this:

```YAML
  bms1: !include
    file: packages/bms/bms_combine_DEYE_CAN_module.yaml
    vars:
      # CANBUS ID
      canbus_id: 'canbus_deye'
      # BMS vars
      bms_id: '1' # must be a number
      bms_name: 'DEYE-BMS 1'

  bms2: !include
    file: packages/bms/bms_combine_DEYE_CAN_module.yaml
    vars:
      # CANBUS ID
      canbus_id: 'canbus_deye'
      # BMS vars
      bms_id: '2' # must be a number
      bms_name: 'DEYE-BMS 2'
```

Regarding the modules CAN id you can do like this with a maximum of 9 modules:
Are there any DIP switches on the module ?
The first module is 0 = can id 0x110 right ?
Do you know how the modules are numbered?

```YAML
- can_id: 0x11${bms_id}
```

I haven't checked everything but here is the information we need to combine.
Some of this information can be generated via template sensors.
I will add SOH to this list.

Tell me the information you can't get from the module.

- bms${bms_id}_online_status
- bms${bms_id}_charging_allowed
- bms${bms_id}_discharging_allowed
- bms${bms_id}_equalizing
- bms${bms_id}_total_voltage
- bms${bms_id}_current
- bms${bms_id}_power
- bms${bms_id}_state_of_charge
- bms${bms_id}_battery_capacity
- bms${bms_id}_capacity_remaining_ah
- bms${bms_id}_charging_cycles
- bms${bms_id}_max_charge_current
- bms${bms_id}_max_discharge_current
- bms${bms_id}_max_cell_voltage
- bms${bms_id}_max_voltage_cell
- bms${bms_id}_min_cell_voltage
- bms${bms_id}_min_voltage_cell
- bms${bms_id}_min_temperature
- bms${bms_id}_min_temperature_sensor
- bms${bms_id}_max_temperature
- bms${bms_id}_max_temperature_sensor
- bms${bms_id}_cell_ovp
- bms${bms_id}_cell_uvp
- bms${bms_id}_balance_trigger_voltage
- bms${bms_id}_errors_bitmask
- bms${bms_id}_charging_power
- bms${bms_id}_discharging_power

GHswitt commented:
This is the battery I'm testing with:

![image](https://github.com/user-attachments/assets/c7ab2dd1-c7e8-40c0-9096-d652211259bc)

It has one top main (BMS) part and two modules. The modules itself do not have any dip switches or other things you could access. You can have up to 5 modules in one unit. If you want more you need to use another main/top unit and connect it with the others via the IN/OUT (InterCAN) ports. The top module has a port called 'PCS' which is normally connected to the inverter.

Up to now I'm only able to get the 16 cell voltages from the IN/OUT port. All other information (combination created from the top module and each module) I get from the 'PCS' port. If you have information on how to get everything from the InterCAN ports I'll take a look at it.

If I understood it correctly you would prefer to create one BMS for each module instead of the current architecture which only provides one BMS, regardless of how many modules are installed:

BMS 1 (main module)
 -> Module 1
 -> Module 2
 -> Module x

Regarding the can_id, yes that looks easy. However I would not like to use the bms_id. It would require that the first BMS is the Deye and this BMS starting with 0. If there are other BMS installed before it wouldn't work.

`- can_id: 0x11${bms_id}`

- bms${bms_id}_online_status: Not yet. Will use a timer which sets it to offline if no frame is received after x seconds
- bms${bms_id}_charging_allowed: Ok
- bms${bms_id}_discharging_allowed: Ok
- bms${bms_id}_equalizing: Ok
- bms${bms_id}_total_voltage: Ok
- bms${bms_id}_current: Ok
- bms${bms_id}_power: Ok (calculated from voltage and current)
- bms${bms_id}_state_of_charge: Ok
- bms${bms_id}_battery_capacity: Ok
- bms${bms_id}_capacity_remaining_ah: Ok (calculated from SoC and capacity)
- bms${bms_id}_charging_cycles: Ok
- bms${bms_id}_max_charge_current: Ok
- bms${bms_id}_max_discharge_current: Ok
- bms${bms_id}_max_cell_voltage: Ok
- bms${bms_id}_max_voltage_cell: **No. Can be calculated together with InterCAN messages.**
- bms${bms_id}_min_cell_voltage: Ok
- bms${bms_id}_min_voltage_cell: **No. Can be calculated together with InterCAN messages.**
- bms${bms_id}_min_temperature: Ok
- bms${bms_id}_min_temperature_sensor: Ok. It only provides one (battery) value.
- bms${bms_id}_max_temperature: Ok
- bms${bms_id}_max_temperature_sensor: Ok. It only provides one (battery) value.
- bms${bms_id}_cell_ovp: **Voltage unknown**
- bms${bms_id}_cell_uvp: **Voltage unknown**
- bms${bms_id}_balance_trigger_voltage: **Voltage unknown**
- bms${bms_id}_errors_bitmask: Ok
- bms${bms_id}_charging_power: Ok (calculated)
- bms${bms_id}_discharging_power: Ok (calculated)

Sleeper85 commented:
I didn't know that the data for each module could be obtained from the PCS port.
This is different from what's explained on [Psynosaur's GitHub](https://github.com/Psynosaur/esphome-deye-bms).

Is this the case for other Deye battery models ?

It's clear that your two-modules battery has two BMS.
I find it more logical to declare both BMS in the main YAML because someone else might have one module or three.
Your YAML with two modules only works for two modules.

You can do this:

```YAML
  bms1: !include
    file: packages/bms/bms_combine_DEYE_CAN_battery_module.yaml
    vars:
      # CANBUS ID
      canbus_id: 'canbus_deye'
      # BMS vars
      bms_id: '1' # must be a number
      bms_name: 'DEYE-BMS 1'
      bms_module_id: '0'

  bms2: !include
    file: packages/bms/bms_combine_DEYE_CAN_battery_module.yaml
    vars:
      # CANBUS ID
      canbus_id: 'canbus_deye'
      # BMS vars
      bms_id: '2' # must be a number
      bms_name: 'DEYE-BMS 2'
      bms_module_id: '1'
```

```YAML
- can_id: 0x11${bms_module_id}
```

Regarding the `canbus_deye` you need to create an additional CAN bus without modifying the existing CAN bus `canbus_node1` which is used for the connection to the inverter.

GHswitt commented:
If you look at his code, only the cell voltages are taken from the InterCAN port:

https://github.com/Psynosaur/esphome-deye-bms/blob/c87e947fe9d0cd0894af4b8439ca0e2d2da72015/deye-can-bus.yaml#L68

For the PCS port, there are fixed ID frames of the main module (summary of all modules) and a number of frames for each pack/module:

https://github.com/Psynosaur/esphome-deye-bms/blob/c87e947fe9d0cd0894af4b8439ca0e2d2da72015/deye-can-bus.yaml#L355

I've seen that it seems possible to get the settings via InterCAN. But I won't get near anywhere sending frames to the battery if this is required to read settings. That would be to dangerous for me, I would stick to listening only.

I like the idea to create one BMS for each module and will change that.

As for the `canbus_deye/canbus_node1`, this was for testing only and will not be present in the final version. In the example I'll create a second canbus based on an SPI controller.

Sleeper85 commented:
It's boring to have to create two different CAN bus just to get the voltages for each cell.
First, focus on the PCS port only.

Regarding the temperature sensors number, you already know the minimum/maximum temperatures for each module, but not the sensor number.

You can create templates sensors and number the sensors yourself just to get the module number information.

**BMS 1 / Module 0**
min_temperature_sensor = (${bms_id} * 100 + ${deye_module_id}1); = 101
max_temperature_sensor = (${bms_id} * 100 + ${deye_module_id}2); = 102

**BMS 2 / Module 1**
min_temperature_sensor = (${bms_id} * 100 + ${deye_module_id}1); = 211
max_temperature_sensor = (${bms_id} * 100 + ${deye_module_id}2); = 212

Sleeper85 commented:
> bms${bms_id}_online_status: Not yet. Will use a timer which sets it to offline if no frame is received after x seconds

```YAML
binary_sensor:
  # Required sensors cannot be retrieved from BMS
  # online_status
  - platform: template
    id: bms${bms_id}_online_status
    name: "${name} ${bms_name} online status"
    lambda: |-
      if (id(bms${bms_id}_total_voltage).state > 0)
        return true;
      else
        return false;
```

GHswitt commented:
I've updated the code, it now creates one BMS for each module.

Will change the example with two CAN ports after I've modified some hardware and tested it with both ports (1x to inverter, 1x to battery PCS port).

Sleeper85 commented:
Your example file should now be called `multi-bms_DEYE-PCS_CAN.yaml`.

I'm going to change the organization of `board.yaml` to make the management of `in/out` interfaces more flexible.

This will allow you to import a board without an interface and then add only the necessary UART/CANBUS interfaces.

In your case, with an AtomS3, the `canbus_inverter` will use GPIOs 5 and 6 (CAN base) and the `canbus_bms` will use GPIOs 2 and 1 (CAN unit).

Sleeper85 commented:
Are you going to add an MCP2515 to your AtomS3 to make two different CAN bus ?

Sleeper85 commented:
[New board options](https://diysolarforum.com/threads/yambms-jk-bms-can-with-new-cut-off-charging-logic-open-source.79325/page-125#post-1400333)

GHswitt commented:
The AtomS3 was only for testing. I've switched to a LilyGo T-CAN485 and added a MCP2515 as secondary CAN port. The example file multi-bms_DEYE-PCS_CAN.yaml was updated and I can confirm this works with my Deye 12K inverter with the settings "Automatic" and PylonV2.

I've also added _minimal versions with far less sensors.

The batteries were full (and the system exporting 11kW to the grid). The YamBMS charging status was 'Float' and the requested voltage 53.6V. This caused a slow discharge of the batteries. Before the current was 0A for some time, requested voltage before was 55.2V from a JK-BMS.

![image](https://github.com/user-attachments/assets/d88d83e4-7ca5-4f75-ad34-568c21aa0b79)

After trying bulk charge the charge voltage did go up as expected:

![image](https://github.com/user-attachments/assets/efe024a4-286d-4cec-9a6e-a62065273caa)

Side note: The Deye battery itself requests a charge voltage of 58.4V, which I find quite high. On the plus side it uses a dynamic charge current compared to the JK-BMS which always provide a fixed charge current.

I will look into the new board options. Next steps for me are:
- Add support for RS485 node: LilyGo T-CAN485 connected to Deye battery and RS485 for YamBMS
- Add support for LilyGo T-Connect Pro

This is my planned setup:

```
Deye battery <- CAN ->   T-CAN485 <->  RS485>|
JK-BMS       <- RS485 -> ESP32 <->     RS485>|
JK-BMS       <----^                          |
Basen-BMS    <- RS485 -> ESP32 <->     RS485>|
Basen-BMS    <----^                          |
Victron SmartShunt 1 |                       |
Victron SmartShunt 2 |                       |
Victron SmartShunt 3 |   -> RP2040 <-> RS485>|
Victron SmartShunt 4 |                       |  
Victron SmartShunt 5 |                       |> T-Connect Pro (YamBMS main) <-> CAN <-> Deye inverter
```

Sleeper85 commented:
@GHswitt 

I should have finished reorganizing the code this evening regarding the `board.yaml` import.

Here are the GPIOs I chose for `LilyGo T-Connect`.

```YAML
substitutions:
  board_chip: "ESP32-S3"
  board_name: "LilyGo T-Connect"
  # uart_esp_1
  tx_pin_1: '4'
  rx_pin_1: '5'
  # uart_esp_2
  tx_pin_2: '6'
  rx_pin_2: '7'
  # uart_esp_3
  tx_pin_3: '17'
  rx_pin_3: '18'
  # canbus_esp32_can
  tx_pin_4: '9'
  rx_pin_4: '10'
  # canbus_mcp2515
  spi_mosi_pin: '11'
  spi_miso_pin: '13'
  spi_clk_pin: '12'
  mcp2515_cs_pin: '14'
```

In the `main.yaml`

```YAML
packages:

  ############### >>> UNCOMMENT YOUR BOARD <<< ###############
  ############### >>>   ONLY ONE BOARD !   <<< ###############

  # This list is incomplete, other board.yaml are available in the board folder

  # device_board: !include packages/board/board_ESP32_Generic.yaml
  # device_board: !include packages/board/board_ESP32_DevKit-V1.yaml
  device_board: !include packages/board/board_ESP32_LilyGo-T-CAN485.yaml
  # device_board: !include packages/board/board_ESP32-C3_ETH01-EVO.yaml
  # device_board: !include packages/board/board_ESP32-S3_DevKitC-1.yaml
  # device_board: !include packages/board/board_ESP32-S3_LilyGo-T-Connect.yaml
  # device_board: !include packages/board/board_ESP32-S3_AtomS3-Lite.yaml
  
  # device_board: !include
  #   file: packages/board/board_ESP32-S3_AtomS3.yaml # board_ESP32-S3_AtomS3R.yaml
  #   vars:
  #     display_auto_next_page_interval: '5s'

  ########### >>> UNCOMMENT YOUR BOARD OPTIONS <<< ###########

  uart_esp_1: !include packages/board/board_options_uart_esp_1.yaml
  # uart_esp_2: !include packages/board/board_options_uart_esp_2.yaml
  # uart_esp_3: !include packages/board/board_options_uart_esp_3.yaml

  # ESP32_CAN
  canbus_node_1: !include
    file: packages/board/board_options_canbus_esp32_can.yaml
    vars:
      canbus_node_id: 'canbus_inverter_1'

  # MCP2515
  canbus_node_2: !include
    file: packages/board/board_options_canbus_mcp2515.yaml
    vars:
      canbus_node_id: 'canbus_bms_1'

  ############################################################
```

Sleeper85 commented:
@GHswitt 

> This is my planned setup:
>```
>Deye battery <- CAN ->   T-CAN485 <->  RS485>|
>JK-BMS       <- RS485 -> ESP32 <->     RS485>|
>JK-BMS       <----^                          |
>Basen-BMS    <- RS485 -> ESP32 <->     RS485>|
>Basen-BMS    <----^                          |
>Victron SmartShunt 1 |                       |
>Victron SmartShunt 2 |                       |
>Victron SmartShunt 3 |   -> RP2040 <-> RS485>|
>Victron SmartShunt 4 |                       |  
>Victron SmartShunt 5 |                       |> T-Connect Pro (YamBMS main) <-> CAN <-> Deye inverter
>```

It looks cool, when it's finished take some nice pictures to show it to the community.

Sleeper85 commented:
[YamBMS 1.5.5 is released !](https://diysolarforum.com/threads/yambms-jk-bms-can-with-new-cut-off-charging-logic-open-source.79325/page-125#post-1401909)

GHswitt commented:
I've updated the branch to match the new features.

Sleeper85 commented:
Thank you for this integration.

Can you check the following points : 

1) Date, Version and GitHub
Change the date and version at the beginning of each YAML file, new YAML files start with version `1.1.1`.
For **YamBMS_LP_DEYE-PCS_CAN.yaml** keep `1.5.5` but change the date.
For **board_options_itf_canbus_mcp2515.yaml** change the date and version to `1.1.2`.
You can use your GitHub address for the YAML files you create.

2) YAML mames, I would like it like this:
* bms_can_DEYE_module_full.yaml => bms_sensors_DEYE_CAN_module_full.yaml
* bms_can_DEYE_module_minimal.yaml => bms_sensors_DEYE_CAN_module_minimal.yaml
* bms_can_DEYE_top_full.yaml => bms_sensors_DEYE_CAN_top_full.yaml
* bms_can_DEYE_top_minimal.yaml => bms_sensors_DEYE_CAN_top_minimal.yaml
* bms_combine_DEYE_CAN_full.yaml => bms_combine_DEYE_CAN_module_full.yaml
* bms_combine_DEYE_CAN_minimal.yaml => bms_combine_DEYE_CAN_module_minimal.yaml

As soon as it's done I'll merge this PR.

GHswitt commented:
Done, thank you. I hope I got it all.

Sleeper85 commented:
At the beginning of each DEYE YAML, on line 3, can you replace the Github address with your Github address?

GHswitt commented:
Done.
